File: firebase/functions/lib/community/on_community_membership.dart
import 'dart:async';

import 'package:enum_to_string/enum_to_string.dart';
import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../utils/infra/firestore_event_function.dart';
import '../utils/infra/on_firestore_helper.dart';
import '../on_firestore_function.dart';
import '../utils/infra/firestore_utils.dart';
import 'package:data_models/community/community.dart';
import 'package:data_models/community/membership.dart';

class OnCommunityMembership extends OnFirestoreFunction<Membership> {
  OnCommunityMembership()
      : super(
          [
            AppFirestoreFunctionData(
              'CommunityMembershipOnCreate',
              FirestoreEventType.onCreate,
            ),
          ],
          (snapshot) {
            return Membership.fromJson(
              firestoreUtils.fromFirestoreJson(snapshot.data.toMap()),
            );
          },
        );

  @override
  String get documentPath =>
      'memberships/{membershipId}/community-membership/{communityMembershipId}';

  @override
  Future<void> onUpdate(
    Change<DocumentSnapshot> changes,
    Membership before,
    Membership after,
    DateTime updateTime,
    EventContext context,
  ) async {
    throw UnimplementedError();
  }

  @override
  Future<void> onCreate(
    DocumentSnapshot documentSnapshot,
    Membership parsedData,
    DateTime updateTime,
    EventContext context,
  ) async {
    print(
      'Community Membership (${documentSnapshot.documentID}) has been created',
    );

    await updateOnboardingSteps(parsedData, documentSnapshot);
  }

  @override
  Future<void> onDelete(
    DocumentSnapshot documentSnapshot,
    Membership parsedData,
    DateTime updateTime,
    EventContext context,
  ) {
    throw UnimplementedError();
  }

  @override
  Future<void> onWrite(
    Change<DocumentSnapshot> changes,
    Membership before,
    Membership after,
    DateTime updateTime,
    EventContext context,
  ) {
    throw UnimplementedError();
  }

  Future<void> updateOnboardingSteps(
    Membership membership,
    DocumentSnapshot originalDocumentSnapshot,
  ) async {
    final communityId = membership.communityId;
    const onboardingStep = OnboardingStep.inviteSomeone;
    final communityDoc =
        firestoreHelper.getPathToCommunityDocument(communityId: communityId);
    final communitySnap = await firestore.document(communityDoc).get();
    if (!communitySnap.exists) {
      print('Community ($communityId) does not exist');
      return;
    }

    if (!communitySnap.data.has(Community.kFieldOnboardingSteps)) {
      print(
        'Community ($communityId) has no field ${Community.kFieldOnboardingSteps}',
      );
      return;
    }
    final List<dynamic> onboardingSteps =
        communitySnap.data.toMap()[Community.kFieldOnboardingSteps];
    final stringOnboardingStep = EnumToString.convertToString(onboardingStep);

    if (onboardingSteps.any((element) => element == stringOnboardingStep)) {
      print(
        '$onboardingStep is already in Community ($communityId) ${Community.kFieldOnboardingSteps}',
      );
      return;
    }

    final communityMemberships = await firestore
        .collectionGroup(FirestoreHelper.kCommunityMemberships)
        .where(Membership.kFieldCommunityId, isEqualTo: communityId)
        .get();
    final amIInCommunity = communityMemberships.documents.any(
      (documentSnapshot) =>
          documentSnapshot.data.getString(Membership.kFieldUserId) ==
          membership.userId,
    );
    if (!amIInCommunity) {
      print(
        'User (${membership.userId}) is not in community ($communityId). Something is wrong',
      );
      return;
    }

    if (!communitySnap.data.has(Community.kFieldCreatorId)) {
      print(
        'Community ($communityId) has no field ${Community.kFieldCreatorId}',
      );
      return;
    }
    final communityCreatorId =
        communitySnap.data.getString(Community.kFieldCreatorId);
    if (communityCreatorId == membership.userId) {
      print(
        'User (${membership.userId}) is a creator of the Community ($communityId)',
      );
      return;
    }

    // Update Community with new step
    final updateMap = {
      Community.kFieldOnboardingSteps: [
        ...onboardingSteps,
        stringOnboardingStep,
      ],
    };

    await communitySnap.reference.updateData(UpdateData.fromMap(updateMap));

    print(
      '$onboardingStep has been added to Community ($communityId) ${Community.kFieldOnboardingSteps}',
    );
  }
}

File: firebase/functions/src/community/on_community_membership.ts
import * as admin from 'firebase-admin'
import { firestore, firestoreUtils } from '../utils/infra/firestore_utils'
import {
    OnFirestoreFunction,
    AppFirestoreFunctionData,
    FirestoreEventType,
} from '../on_firestore_function'
import { Membership, MembershipStatus, Community, OnboardingStep } from '../types'
import { FirestoreHelper } from '../utils/infra/on_firestore_helper'
import { membershipIsMember } from '../types'

export class OnCommunityMembership extends OnFirestoreFunction<Membership> {
    constructor() {
        super(
            [
                {
                    functionName: 'CommunityMembershipOnCreate',
                    firestoreEventType: FirestoreEventType.onCreate,
                },
            ],
            (snapshot) => {
                return firestoreUtils.fromFirestoreJson(
                    snapshot.data() ?? {}
                ) as unknown as Membership
            }
        )
    }

    get documentPath(): string {
        return 'memberships/{membershipId}/community-membership/{communityMembershipId}'
    }

    async onCreate(
        documentSnapshot: admin.firestore.DocumentSnapshot,
        parsedData: Membership,
        _updateTime: Date,
        _context: import('firebase-functions').EventContext
    ): Promise<void> {
        console.log(`Community Membership (${documentSnapshot.id}) has been created`)
        await this._updateOnboardingSteps(parsedData, documentSnapshot)
    }

    private async _updateOnboardingSteps(
        membership: Membership,
        originalSnapshot: admin.firestore.DocumentSnapshot
    ): Promise<void> {
        const communityId = membership.communityId
        const communityDoc = await firestore
            .doc(new FirestoreHelper().getPathToCommunityDocument({ communityId }))
            .get()

        if (!communityDoc.exists) {
            console.log(`Community (${communityId}) does not exist`)
            return
        }

        const communityData = communityDoc.data() as Community
        const onboardingSteps: string[] = communityData.onboardingSteps ?? []
        const step = OnboardingStep.inviteSomeone.toString()

        if (onboardingSteps.includes(step)) {
            console.log(`${step} already in Community (${communityId}) onboardingSteps`)
            return
        }

        // Check if user is community creator
        const creatorId = communityData.creatorId
        if (creatorId === membership.userId) {
            console.log(
                `User (${membership.userId}) is a creator of the Community (${communityId})`
            )
            return
        }

        // Verify user is in community
        const communityMemberships = await firestore
            .collectionGroup(FirestoreHelper.kCommunityMemberships)
            .where('communityId', '==', communityId)
            .get()

        const isUserInCommunity = communityMemberships.docs.some(
            (doc) => (doc.data() as Membership).userId === membership.userId
        )

        if (!isUserInCommunity) {
            console.log(
                `User (${membership.userId}) is not in community (${communityId}). Something is wrong`
            )
            return
        }

        await communityDoc.ref.update({ onboardingSteps: [...onboardingSteps, step] })
        console.log(`${step} has been added to Community (${communityId}) onboardingSteps`)
    }
}
